### ISSUE TITLE
Implement RNN

### ISSUE THREAD

**Issue by gtmx23** [coruubc/diverse_price_prediction#55]
- Implement RNN (pytorch for ability to explain what we are doing)
- Tune RNN
- Documentation

**Comment by zth96** [comment]
Hello here are the updates
- Defined Simple Recurrent Neural Network with 128 hidden units in RNN layer and 2 fully connected layers
- Also tried Recurrent Neural Network with an LSTM layer with 128 hidden units, one LSTM layer and 2 fully connected layers using ReLu activation
-  Tunning for Basic RNN and  Tunning for RNN with LSTM layer
Yet I still have the issue with kernel crushing. I tried on my other computer as well and it works alright when I first run it but the rest of the times it just keeps crushing; so as of now I do not have the results but the codes I believe are clean and ready.
Here is a general summary of our work that has been done on branch 44 under neural_network.ipynb file.
https://github.com/coruubc/diverse_price_prediction/blob/44-build-custom-neural-network/neural_network.ipynb
Load and preprocess data
RMSE and MAE calculation functions
Define preprocessor
Set up expanding window: 1st trainset 1990-2010, 1st validset 2011-2016
Defined Simple Layer Neural Network (basic neural network with two hidden layers containing 128 and 64 neurons,)
Defined Deep Nueral Network model with three hidden layers (256, 128, and 64 neurons) and additional regularization techniques
Defined Simple Recurrent Neural Network with 128 hidden units in RNN layer and 2 fully connected layers
Defined LSTM Recurrent Neural Network with 128 hidden units, one LSTM layer and 2 fully connected layers using ReLu activation
Created functions to compute the root mean squared error (RMSE) and mean absolute error (MAE)
Expanding window setup starting from 1990 to 2001 for training and 2002 to 2006 for validation
Formulate data preprocessing in a loop and model training and validation with an Adam optimizer and MSE loss
Hyperparameter tuning with a grid search over learning rates and batch sizes to find the optimal combination, including an early stopping mechanism to halt training
Tunning for Basic NN
Tunning for Basic RNN
Tunning for RNN with LSTM layer

**Comment by zth96** [comment]
Updates of today. 
-  Load the data frame and display the first few rows. 
- Drop the outliers
- Checking data type, preprocess the data and define target and features. 
- Classify the features and define preprocessors
- Defined Neural Network Model to mimic ELM
- Define Simple Neural Network Model
- Define a deep Neural Network Model
- Define deeper Neural Network Model
- Define a deep Neural Network Model with a large 1st hidden layer for pattern recognision
- Define simple basic Recurrent Neural Network model
- Define RNN with multiple layers
- Define RNN model with regularization techniques/dropout
- Defined bidirectional RNN (both forward and backward pass)
- Defined a hybrid of CNN and RNN model
- RMSE and MAE calculation functions
- Expanding window CV codes (running in to problem, will work on it later to have test for 2007 to 2010) -- pending documentation
- Tunning for Neural network models
- Tunning for RNN models (pending todo)
- Output results and time

**Comment by zth96** [comment]
The MultiLayerRNN generally performs better in terms of RMSE and MAE compared to the SimpleRNN and RegularizedRNN, suggesting more complex or deeper models might be capturing the patterns in the data more effectively.

The BiDirectionalRNN shows competitive results, indicating that accessing past and future context simultaneously can improve performance.

The RegularizedRNN, despite having dropout, does not necessarily perform better than other models. This might indicate that either the dropout is not set optimally, or the data does not have substantial overfitting issues that dropout would mitigate.

Next step would be tunning multilayer RNN which is the best performing one for now.

**Comment by zth96** [comment]
Updated RNN with average CV score through 5 folds and test score. 

<img width="542" alt="image" src="https://github.com/coruubc/diverse_price_prediction/assets/143483779/67a719b5-afc0-4c0b-90b1-e306692b80a0">

SimpleRNN Analysis:
The SimpleRNN model has higher error rates both during cross-validation (Average RMSE = 8.46, Average MAE = 5.11) and on the external test dataset (Test RMSE = 11.88, Test MAE = 6.43). This pattern suggests that the basic RNN configuration may not effectively capture the more intricate patterns within the dataset, potentially due to its simplicity.

MultiLayerRNN Analysis:
The MultiLayerRNN shows an improvement in average RMSE during cross-validation (7.72). However, its performance dropped significantly on the test dataset, recording the highest RMSE of 12.30. The discrepancy likely points to overfitting; the model appears to fit too closely to the training data, losing its ability to generalize well when faced with new, unseen data.

RegularizedRNN Analysis:
Incorporating regularization techniques such as dropout to mitigate overfitting. It posts slightly poorer cross-validation scores compared to the MultiLayerRNN yet achieves a better test RMSE (11.92). This improvement in test performance suggests that the RegularizedRNN may be better to generalize, avoiding some of the overfitting issues seen in the MultiLayerRNN.

BiDirectionalRNN Analysis: [Best performing one]
The BiDirectionalRNN stands out with the #strongest# performance balance between validation and testing phases, achieving the lowest test RMSE (11.57) of all the models tested. This performance indicates that processing data in both forward and backward directions allows the model to better understand and utilize the contextual relationships within the data, enhancing its generalization capabilities.

SimpleLSTM Analysis: 
The SimpleLSTM model demonstrates competitive error rates during cross-validation with an Average RMSE of 8.34 and Average MAE of 4.82, which indicates a robust performance during training phases. On the test dataset, the model records an RMSE of 12.04 and an MAE of 6.16. This suggests that while the LSTM's capability to capture and remember long-term dependencies is beneficial, there is still a noticeable gap between training and testing performance, indicating a potential for slight overfitting or inadequate capture of more complex patterns in the test data.
